File: 0022-generate-parentheses/0022-generate-parentheses.cpp
class Solution
{
public:
    vector<string> generateParenthesis(int n)
    {
        vector<char> path;
        vector<string> res;
        dfs(0, path, 0, 0, res, n);
        return res;
    }
    void dfs(int startIndex, vector<char> &path, int openCount, int closeCount, vector<string> &res, int n)
    {
        if (path.size() == 2 * n)
        {
            res.emplace_back(path.begin(), path.end());
            return;
        }
        if (openCount < n)
        {
            path.emplace_back('(');
            dfs(startIndex + 1, path, openCount + 1, closeCount, res, n);
            path.pop_back();
        }
        if (closeCount < openCount)
        {
            path.emplace_back(')');
            dfs(startIndex + 1, path, openCount, closeCount + 1, res, n);
            path.pop_back();
        }
    }
};

File: 0022-generate-parentheses/0022-generate-parentheses.js
function generateParenthesis(n) {
  function dfs(startIndex, path, openCount, closeCount, res) {
    if (startIndex == 2 * n) {
      // make a deep copy since otherwise we'd be append the same list over and over
      res.push(path.join(""));
      return;
    }
    if (openCount < n) {
      path.push("(");
      dfs(startIndex + 1, path, openCount + 1, closeCount, res);
      path.pop();
    }
    if (closeCount < openCount) {
      path.push(")");
      dfs(startIndex + 1, path, openCount, closeCount + 1, res);
      path.pop();
    }
  }

  const res = [];
  dfs(0, [], 0, 0, res);
  return res;
}